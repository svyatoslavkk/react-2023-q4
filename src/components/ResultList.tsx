import React from 'react';
import { useRouter } from 'next/router';
import Image from 'next/image';
import { updateQueryParams } from '../functions/updateQueryParams';
import { Character } from '../interfaces/interfaces';

function ResultList(props: Record<'searchResults', readonly Character[]>) {
  const imgUrl =
    'https://mir-s3-cdn-cf.behance.net/project_modules/1400/010cc167400543.5b38baa438d6f.png';
  const { searchResults } = props;
  const router = useRouter();

  return (
    <>
      <div className="result-list" data-testid="search-results">
        {searchResults.length ? (
          <div className="fetching-results">
            {searchResults.map((character: Character) => (
              <div
                key={character.uid}
                className="result-list-item"
                data-testid="card"
                onClick={() => {
                  router.push(
                    '?' +
                      updateQueryParams(
                        router.query,
                        'details',
                        character.name,
                      ),
                  );
                }}
              >
                <Image
                  className="result-list-item-image"
                  src={imgUrl}
                  alt="Character Image"
                  width={50}
                  height={50}
                />
                <div className="result-list-item-text-content">
                  <h3 className="result-list-item-name">{character.name}</h3>
                </div>
              </div>
            ))}
          </div>
        ) : (
          <p data-testid="no-found">No results found.</p>
        )}
      </div>
    </>
  );
}

export default ResultList;
