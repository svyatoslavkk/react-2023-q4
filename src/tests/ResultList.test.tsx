import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { createCardsListResponseMock } from './mocks/_mocks';
import ResultList from '../components/ResultList';

vi.mock('next/router', () => vi.importActual('next-router-mock'));

describe('Tests for the Card List component', () => {
  it('Component renders the specified number of cards - 7', async () => {
    render(
      <ResultList searchResults={createCardsListResponseMock(7, 10, false)} />,
    );

    expect(await screen.findAllByText(/testCard/)).toHaveLength(7);
  });

  it('An appropriate message is displayed if no cards are present', async () => {
    render(<ResultList searchResults={[]} />);

    expect(screen.getByText('No results found.')).toBeTruthy();

    expect(screen.queryByText(/testCard/)).toBeFalsy();
  });
});
