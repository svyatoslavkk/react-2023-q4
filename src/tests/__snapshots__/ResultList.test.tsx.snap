// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Card List component > An appropriate message is displayed if no cards are present 1`] = `
<body>
  <div>
    <div
      class="result-list"
    >
      <p>
        No results found.
      </p>
    </div>
  </div>
</body>
`;

exports[`Tests for the Card List component > Component renders the specified number of cards - 7 1`] = `
<body>
  <div>
    <div
      class="result-list"
    >
      <ul
        class="fetching-results"
      >
        <li
          class="result-list-item"
        >
          <img
            alt="Character Image"
            class="result-list-item-image"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
          <div
            class="result-list-item-text-content"
          >
            <h3
              class="result-list-item-name"
            >
              testCard-0
            </h3>
          </div>
        </li>
        <li
          class="result-list-item"
        >
          <img
            alt="Character Image"
            class="result-list-item-image"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
          <div
            class="result-list-item-text-content"
          >
            <h3
              class="result-list-item-name"
            >
              testCard-1
            </h3>
          </div>
        </li>
        <li
          class="result-list-item"
        >
          <img
            alt="Character Image"
            class="result-list-item-image"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
          <div
            class="result-list-item-text-content"
          >
            <h3
              class="result-list-item-name"
            >
              testCard-2
            </h3>
          </div>
        </li>
        <li
          class="result-list-item"
        >
          <img
            alt="Character Image"
            class="result-list-item-image"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
          <div
            class="result-list-item-text-content"
          >
            <h3
              class="result-list-item-name"
            >
              testCard-3
            </h3>
          </div>
        </li>
        <li
          class="result-list-item"
        >
          <img
            alt="Character Image"
            class="result-list-item-image"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
          <div
            class="result-list-item-text-content"
          >
            <h3
              class="result-list-item-name"
            >
              testCard-4
            </h3>
          </div>
        </li>
        <li
          class="result-list-item"
        >
          <img
            alt="Character Image"
            class="result-list-item-image"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
          <div
            class="result-list-item-text-content"
          >
            <h3
              class="result-list-item-name"
            >
              testCard-5
            </h3>
          </div>
        </li>
        <li
          class="result-list-item"
        >
          <img
            alt="Character Image"
            class="result-list-item-image"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fmir-s3-cdn-cf.behance.net%2Fproject_modules%2F1400%2F010cc167400543.5b38baa438d6f.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
          <div
            class="result-list-item-text-content"
          >
            <h3
              class="result-list-item-name"
            >
              testCard-6
            </h3>
          </div>
        </li>
      </ul>
    </div>
  </div>
</body>
`;
